syntax = "proto3";

import "bess_message.proto";
import "module_message.proto";
package bess.protobuf;

service BESSControl {
  rpc ResetAll (Empty) returns (EmptyResponse) {}

  rpc PauseAll (Empty) returns (EmptyResponse) {}
  rpc ResumeAll (Empty) returns (EmptyResponse) {}

  rpc ResetWorkers (Empty) returns (EmptyResponse) {}
  rpc ListWorkers (Empty) returns (ListWorkersResponse) {}
  rpc AddWorker (AddWorkerRequest) returns (EmptyResponse) {}
  // TODO: delete_worker

  rpc ResetTcs (Empty) returns (EmptyResponse) {}
  rpc ListTcs (ListTcsRequest) returns (ListTcsResponse) {}
  rpc AddTc (AddTcRequest) returns (EmptyResponse) {}
  rpc GetTcStats (GetTcStatsRequest) returns (GetTcStatsResponse) {}

  rpc ListDrivers (Empty) returns (ListDriversResponse) {}
  rpc GetDriverInfo(GetDriverInfoRequest) returns (GetDriverInfoResponse) {}
  // TODO: rpc import_driver()

  rpc ResetPorts (Empty) returns (EmptyResponse) {}
  rpc ListPorts (Empty) returns (ListPortsResponse) {}
  rpc CreatePort (CreatePortRequest) returns (CreatePortResponse) {}
  rpc DestroyPort (DestroyPortRequest) returns (EmptyResponse) {}
  rpc GetPortStats (GetPortStatsRequest) returns (GetPortStatsResponse) {}

  // TODO: mclass

  rpc ResetModules (Empty) returns (EmptyResponse) {}
  rpc ListModules (Empty) returns (ListModulesResponse) {}
  rpc CreateModule (CreateModuleRequest) returns (CreateModuleResponse) {}
  rpc DestroyModule (DestroyModuleRequest) returns (EmptyResponse) {}
  rpc GetModuleInfo (GetModuleInfoRequest) returns (GetModuleInfoResponse) {}
  rpc ConnectModules (ConnectModulesRequest) returns (EmptyResponse) {}
  rpc DisconnectModules (DisconnectModulesRequest) returns (EmptyResponse) {}

  rpc AttachTask (AttachTaskRequest) returns (EmptyResponse) {}

  rpc EnableTcpdump (EnableTcpdumpRequest) returns (EmptyResponse) {}
  rpc DisableTcpdump (DisableTcpdumpRequest) returns (EmptyResponse) {}

  rpc KillBess (Empty) returns (EmptyResponse) {}

  // TODO: modules commands
  rpc BPFCommandAdd (BPFArg) returns (EmptyResponse) {}
  rpc BPFCommandClear (BPFCommandClearArg) returns (EmptyResponse) {}

  rpc DumpCommandSetInterval (DumpArg) returns (EmptyResponse) {}
  rpc ExactMatchCommandAdd (ExactMatchCommandAddArg) returns (EmptyResponse) {}
  rpc ExactMatchCommandDelete (ExactMatchCommandDeleteArg) returns (EmptyResponse) {}
  rpc ExactMatchCommandClear (ExactMatchCommandClearArg) returns (EmptyResponse) {}
  rpc ExactMatchCommandSetDefaultGate (ExactMatchCommandSetDefaultGateArg) returns (EmptyResponse) {}

  rpc HashLBCommandSetMode (HashLBCommandSetModeArg) returns (EmptyResponse) {}
  rpc HashLBCommandSetGates (HashLBCommandSetGatesArg) returns (EmptyResponse) {}

  rpc IPLookupCommandAdd (IPLookupCommandAddArg) returns (EmptyResponse) {}
  rpc IPLookupCommandClear (IPLookupCommandClearArg) returns (EmptyResponse) {}

  rpc L2ForwardCommandAdd (L2ForwardCommandAddArg) returns (EmptyResponse) {}
  rpc L2ForwardCommandDelete (L2ForwardCommandDeleteArg) returns (EmptyResponse) {}
  rpc L2ForwardCommandSetDefaultGate (L2ForwardCommandSetDefaultGateArg) returns (EmptyResponse) {}
  rpc L2ForwardCommandLookup (L2ForwardCommandLookupArg) returns (L2ForwardCommandLookupResponse) {}
  rpc L2ForwardCommandPopulate (L2ForwardCommandPopulateArg) returns (EmptyResponse) {}

  rpc MeasureCommandGetSummary (MeasureCommandGetSummaryArg) returns (MeasureCommandGetSummaryResponse) {}

  rpc PortIncCommandSetBurst (PortIncCommandSetBurstArg) returns (EmptyResponse) {}

  rpc QueueIncCommandSetBurst (QueueIncCommandSetBurstArg) returns (EmptyResponse) {}

  rpc QueueCommandSetBurst (QueueCommandSetBurstArg) returns (EmptyResponse) {}
  rpc QueueCommandSetSize (QueueCommandSetSizeArg) returns (EmptyResponse) {}

  rpc RandomUpdateCommandAdd (RandomUpdateArg) returns (EmptyResponse) {}
  rpc RandomUpdateCommandClear (RandomUpdateCommandClearArg) returns (EmptyResponse) {}

  rpc RewriteCommandAdd (RewriteArg) returns (EmptyResponse) {}
  rpc RewriteCommandClear (RewriteCommandClearArg) returns (EmptyResponse) {}

  rpc RoundRobinCommandSetMode (RoundRobinCommandSetModeArg) returns (EmptyResponse) {}
  rpc RoundRobinCommandSetGates (RoundRobinCommandSetGatesArg) returns (EmptyResponse) {}

  rpc SourceCommandSetBurst (SourceCommandSetBurstArg) returns (EmptyResponse) {}
  rpc SourceCommandSetPktSize (SourceCommandSetPktSizeArg) returns (EmptyResponse) {}

  rpc UpdateCommandAdd (UpdateArg) returns (EmptyResponse) {}
  rpc UpdateCommandClear (UpdateCommandClearArg) returns (EmptyResponse) {}

  rpc VLANPushCommandSetTci (VLANPushArg) returns (EmptyResponse) {}

  rpc WildcardMatchCommandAdd (WildcardMatchCommandAddArg) returns (EmptyResponse) {}
  rpc WildcardMatchCommandDelete (WildcardMatchCommandDeleteArg) returns (EmptyResponse) {}
  rpc WildcardMatchCommandClear (WildcardMatchCommandClearArg) returns (EmptyResponse) {}
  rpc WildcardMatchCommandSetDefaultGate (WildcardMatchCommandSetDefaultGateArg) returns (EmptyResponse) {}

}
